libmisc/salt.c

```diff
@@ -12,19 +12,302 @@
 #ident "$Id$"
 
 #include <sys/time.h>
+#include <crypt.h>
+#include <string.h>
 #include <stdlib.h>
 #include <stdio.h>
 #include <assert.h>
 #include "prototypes.h"
 #include "defines.h"
 #include "getdef.h"
 
-/* local function prototypes */
-static void seedRNG (void);
-static /*@observer@*/const char *gensalt (size_t salt_size);
+#if (defined CRYPT_GENSALT_IMPLEMENTS_AUTO_ENTROPY && \
+     CRYPT_GENSALT_IMPLEMENTS_AUTO_ENTROPY)
+/* Maximum size of the setting for crypt_gensalt(). */
+#define CRYPT_GENSALT_SETTING_SIZE 16
+#define USE_XCRYPT_GENSALT 1
+#else
+#define USE_XCRYPT_GENSALT 0
+#endif
+
+/*
+ * Add the salt prefix.
+ */
+#define MAGNUM(array,ch)	(array)[0]=(array)[2]='$',(array)[1]=(ch),(array)[3]='\0'
+#ifdef USE_BCRYPT
+/*
+ * Using the Prefix $2b$ to enable an anti-collision safety measure in musl libc.
+ * Negatively affects a subset of passwords containing the '\xff' character,
+ * which is not valid UTF-8 (so "unlikely to cause much annoyance").
```

Comment:
When revised to say `$2b$`, this comment became factually wrong. The "anti-collision safety measure" was only enabled for `$2a$`.

Reply:
To be clear, I think this comment should simply be dropped. In fact, it never made sense to deliberately enable the anti-collision safety measure for new password, like we do in the shadow package. That measure was only intended for authentication against old potentially-buggy hashes, not for new password hashes.

Reply:
Fixed in rebased commit.